scripts/build.py

```diff
@@ -631,25 +734,80 @@ def rebuild_new_files(
-        ensure_directory_exists(target_file_path)
-        minify_func(source_file_path, target_file_path, file_hashes, file_name)
+        if not does_directory_tree_exist(target_file_path):
+            os.makedirs(os.path.dirname(target_file_path))
+        task = threading.Thread(target=minify_func, args=(
+            source_file_path, target_file_path, file_hashes, file_name))
+        build_tasks.append(task)
 
 
-def remove_deleted_files(dev_directory, staging_directory):
-    """Walk the staging directory and remove files that have since been removed
-    from the DEV directory.
+def build_directory(source_dir, file_hashes, build_tasks):
+    """Build all files in provided DEV directory if there is no existing staging
+    dir. Otherwise, selectively build recently changed files in
+    said DEV directory and copy to staging dir.
 
     Args:
-        dev_directory: str. Path relative to /oppia where DEV files are located.
+        source_dir: dict(str, str). Directory dict containing directory paths
+        to /DEV dir, staging dir and /build dir .
+        file_hashes: dict(str, str). Dictionary of file hashes.
+        build_tasks: deque(thread). A deque that contains all build tasks queued
+            to be processed.
+    """
+    dev_dir = source_dir.get('dev_dir')
+    staging_dir = source_dir.get('staging_dir')
+    out_dir = source_dir.get('out_dir')
+    delete_tasks = collections.deque()
+    if not os.path.isdir(staging_dir):
+        # If there is no staging dir, perform build process on all files.
+        print 'Creating new %s folder' % staging_dir
+        build_files(dev_dir, staging_dir, file_hashes, build_tasks)
+    else:
+        # If staging dir exists, rebuild all HTML and Python files.
+        mandatory_file_formats = ('.html', '.py')
+        print (
+            'Staging dir exists, re-building all %s files'
+            % str(mandatory_file_formats))
+        build_files(
+            dev_dir, staging_dir, file_hashes, build_tasks,
+            file_formats=mandatory_file_formats)
+        # Compare /DEV with /build dir using file hashes.
+        dev_dir_hashes = get_file_hashes(dev_dir)
+        print "Comparing file hashes of %s and %s" % (dev_dir, out_dir)
+        recently_changed_filenames = get_recently_changed_filenames(
+            dev_dir_hashes, out_dir)
+        if recently_changed_filenames:
+            # Only re-build files that have changed since last build.
+            print "Re-building recently changed files at %s" % dev_dir
+            rebuild_new_files(
+                dev_dir, staging_dir, recently_changed_filenames, file_hashes,
+                build_tasks)
+            # Clean up files in staging dir that have been removed in DEV dir.
+            print 'Scanning directory %s to remove deleted file' % staging_dir
+            remove_deleted_files(dev_dir_hashes, staging_dir, delete_tasks)
+            try:
+                _execute_tasks(delete_tasks)
```

Comment:
A possible idea: perhaps the try/except could go within _execute_tasks. Not sure, just putting it out there for consideration since I see this block of code repeated quite a lot.

Reply:
Yep, I tracked down which statement that could raise an exception and found out it was task.start() and enclose with the try/catch block. Thanks!